theano/tensor/nnet/abstract_conv.py

```diff
@@ -1751,21 +1961,34 @@ def __init__(self, convdim,
             filter_dilation = (1,) * convdim
 
         if isinstance(border_mode, integer_types):
+            if border_mode < 0:
+                raise ValueError(
+                    'invalid border_mode {}, which must be a '
+                    'non-negative integer'.format(border_mode))
             border_mode = (border_mode,) * convdim
-        if isinstance(border_mode, tuple):
+        elif isinstance(border_mode, tuple):
             if len(border_mode) != convdim:
                 raise ValueError(
-                    'border mode must have exactly {} values, '
-                    'but was {}'.format(convdim, border_mode))
-            border_mode = tuple(map(int, border_mode))
-        if border_mode == (0,) * convdim:
-            border_mode = 'valid'
-        if not ((isinstance(border_mode, tuple) and min(border_mode) >= 0) or
-                border_mode in ('valid', 'full', 'half')):
+                    'invalid border_mode {}, which must be a '
+                    'tuple of length {}'.format(border_mode, convdim))
+            for mode in border_mode:
+                if isinstance(mode, tuple) and convdim != 2:
+                    raise NotImplementedError(
+                        'Asymmetric padding not implemented for {}D'.format(convdim))
+                if not((isinstance(mode, integer_types) and mode >= 0) or
+                        (isinstance(mode, tuple) and len(mode) == 2 and min(mode) >= 0 and
+                         all(isinstance(m, integer_types) for m in mode))):
+                    raise ValueError(
+                        'invalid border mode {}. The tuple can only contain integers '
+                        ' or pairs of integers'.format(border_mode))
```

Comment:
Since this is a type of optimization, how about I do this in dnn.py, in the places where we currently return None?

Reply:
If you do it in dnn.py, you will need to do it in many optimization. This would duplicate much code. You could do it in dnn.py, but in the \__init__ method. This would work and won't duplicate code. But I prefer to do it at the earliest. I see this as a simplication. Maybe later other places would need that. So doing it very early would make sure they work directly without needing to handle that case.

But if you do it in the \__init__.py in dnn.py, I would also accept that. Just not as much future proof.